aws/cloudformation/bootstrap_chef_stack.sh.erb

```diff
@@ -76,14 +76,21 @@ CHAIN=$(aws s3 cp $CURRENT/chain.pem - | replace_newline)
 KEY=$(aws s3 cp $CURRENT/key.pem - | replace_newline)
 <% end -%>
 
-<% if database && !rack_env?(:production) -%>
 apt-get install -y  jq
+
+<% if database && !rack_env?(:production) -%>
 DB_SECRET_NAME=${DatabaseSecretAdmin}
 DB_SECRET=$(aws --region $REGION secretsmanager get-secret-value --secret-id $DB_SECRET_NAME --version-stage AWSCURRENT | jq -r ".SecretString")
 DB_USERNAME=$(echo $DB_SECRET | jq -r ".username")
 DB_PASSWORD=$(echo $DB_SECRET | jq -r ".password")
 <% end -%>
 
+
+# Only adhocs have a GitHub token AWS Secret provisioned. Chef Managed environments (staging/test/levelbuilder/production)
+# use an SSH key configured in a Chef Attribute (cdo-github-access.id_rsa, cdo-github-access.id_rsa.pub) that they use to
+# authenticate to GitHub. Adhocs provisioned with `CHEF_MANAGED=true` also have the SSH Chef Attributes.
+GITHUB_TOKEN=$(aws --region ${!REGION} secretsmanager get-secret-value --secret-id "${!ENVIRONMENT}/cdo/github_token" --version-stage AWSCURRENT | jq -r ".SecretString")
```

Comment:
Can you explain (ideally in a comment in the actual code) what the exclamation mark is doing in the parameter expansions of `REGION` and `ENVIRONMENT`? My understanding of bash is that `${!VAR}` represents [indirect expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html#Shell-Parameter-Expansion-1) of `VAR`, but I wouldn't expect us to have environment variables defined that match those values.

Is this actually some other syntax that's a part of one of the other rendering engines this template passes through?

Reply:
This UserData script is enclosed with a CloudFormation template string interpolation function (`!Sub`) which uses the same syntax (`${}`) as bash Braced parameter expansion. `${!` escapes the CloudFormation string interpolation.

We don't really need to use Braced parameter expansion here, so I've switched this to use simple parameter expansion to avoid having to explain it.